### PR TITLE
Add clearance gem for authentication

### PR DESCRIPTION
**Pull request by entcheva**
* remove has_secure_password authentication
* remove custom current_user method
* generate clearance views

### CODE REVIEW

File: app/controllers/users_controller.rb [hunk 1]
```diff
@@ -17,6 +17,6 @@ def create
   private
 
   def user_params
-    params.require(:user).permit(:username, :email, :password, :password_confirmation)
+    params.require(:user).permit(:email, :password)
```
Comment (Euraldius) [inline app/controllers/users_controller.rb]: Let's see if we can implement all of the existing functionality using Clearance, so our theoretical users don't lose the ability to confirm their password and choose a username

File: app/views/users/new.html.erb [hunk 2]
```diff
@@ -1,15 +1,15 @@
-<% if flash[:notice] %>
-  <div class="notice"><%= flash[:notice] %></div>
-<% end %>
+<div id="clearance" class="sign-up">
```
Comment (Euraldius) [inline app/views/users/new.html.erb]: Hi clearance views! Let's keep only the ones of these we need to modify before merging the PR

(I think that will be only `users/_form`

File: app/controllers/users_controller.rb (hunk 1 above)
Comment (entcheva) [inline app/controllers/users_controller.rb]: Hi @Euraldius - based on this [thoughtbot blog post on skipping password confirmation](https://robots.thoughtbot.com/email-and-password-confirmation-bias#no-password-confirmation), as well as our product preference, I added back the username functionality but will skip password confirmation.

File: app/views/passwords/edit.html.erb [hunk 3]
```diff
@@ -0,0 +1,18 @@
+<div id="clearance" class="password-reset">
```
Comment (Euraldius) [inline app/views/passwords/edit.html.erb]: Are the `password/` files generated by clearance? If so I think we can delete them

File: config/locales/clearance.en.yml [hunk 4]
```diff
@@ -0,0 +1,59 @@
+---
```
Comment (Euraldius) [inline config/locales/clearance.en.yml]: Sanity check: is this also a file we can delete?

File: db/migrate/20170509171615_remove_password_digest_from_user.rb [hunk 5]
```diff
@@ -0,0 +1,5 @@
+class RemovePasswordDigestFromUser < ActiveRecord::Migration[5.0]
```
Comment (Euraldius) [inline db/migrate/20170509171615_remove_password_digest_from_user.rb]: oh huh was `password_digest` a bcrypt gem thing?

File: spec/models/snippet_spec.rb [hunk 6]
```diff
@@ -25,8 +25,8 @@
     end
 
     it "has a user_id" do
-      user = FactoryGirl.build_stubbed(:user)
-      snippet = FactoryGirl.build_stubbed(:snippet) 
+      user = FactoryGirl.build_stubbed(:user, id: 2)
```
Comment (Euraldius) [inline spec/models/snippet_spec.rb]: I would write this spec as:

```ruby
it "belongs to a user"
user = FactoryGirl.build_stubbed(:user)
snippet = FactoryGirl.build_stubbed(:snippet, user: user)

expect(snippet.user).to eq(user)
```

It's dangerous to set an id explicitly, since if 🏭  👧 or anything else that's auto-incrementing tries to set the same id, ActiveRecord will throw an error

I recommend you use the [shoulda matchers](https://github.com/thoughtbot/shoulda-matchers) gem to test associations

File: spec/models/user_spec.rb [hunk 7]
```diff
@@ -7,22 +7,17 @@
     it { is_expected.to have_one(:dashboard) }
   end
 
-  describe "has secure password" do
-    it { is_expected.to have_secure_password }
-  end
-
   describe "is valid" do
```
Comment (Euraldius) [inline spec/models/user_spec.rb]: You can use shoulda matchers to test validations as well

File: .gitignore [hunk 8]
```diff
@@ -15,3 +15,5 @@
 
 # Ignore Byebug command history file.
 .byebug_history
+
+.DS_Store
```
Comment (mike-burns) [inline .gitignore]: It's possible to have a git `core.excludesfile` configuration that points to a global gitignore. We do this in our dotfiles, for example:

https://github.com/thoughtbot/dotfiles/blob/master/gitconfig#L16

Which points to a global one:

https://github.com/thoughtbot/dotfiles/blob/master/gitignore

And that's how our macOS users hide their `.DS_Store`.

File: app/controllers/users_controller.rb [hunk 9]
```diff
@@ -1,12 +1,10 @@
...
+class UsersController < Clearance::UsersController
 
   def create
-    @user = User.new(user_params)
+    @user = user_from_params
     if @user.save
-      session[:user_id] = @user.id
+      sign_in @user
+      flash[:notice] = "Sign Up successful! You rock."
       redirect_to dashboard_path
```
Comment (mike-burns) [inline app/controllers/users_controller.rb]: You can combine `flash[:notice]` with `redirect_to`:

```ruby
redirect_to dashboard_path, "Sign up successful! You rock."
```

For more, find for "assign a flash" on http://api.rubyonrails.org/classes/ActionController/Redirecting.html#method-i-redirect_to

File: .gitignore (hunk 8 above)
Comment (georgebrock) [inline .gitignore]: You can set up a global ignore file for things like `.DS_Store` that are going to pop up in every project: https://help.github.com/articles/ignoring-files/#create-a-global-gitignore

File: config/locales/clearance.en.yml (hunk 4 above)
Comment (mike-burns) [inline config/locales/clearance.en.yml]: I _think_ Clearance uses these. Not 100% on that.

File: config/routes.rb [hunk 10]
```diff
@@ -1,11 +1,6 @@
...
   root to: 'dashboards#index', via: :get
   resource :dashboard, only: [:show]
-
-  get '/login', to: 'sessions#new'
-  post '/login', to: 'sessions#create'
-  get '/logout', to: 'sessions#destroy'
-
   resources :snippets, only: [:new, :create]
-  resources :users, only: [:new, :create]
+  resources :users, controller: :users, only: [:create]
```
Comment (mike-burns) [inline config/routes.rb]: Do we need `controller: :users`?

File: db/migrate/20170509171615_remove_password_digest_from_user.rb (hunk 5 above)
Comment (mike-burns) [inline db/migrate/20170509171615_remove_password_digest_from_user.rb]: Clearance calls it `encrypted_password`.

Should this rename the column instead of dropping it?

File: app/controllers/users_controller.rb [hunk 11]
```diff
@@ -16,7 +14,19 @@ def create
...
+  def user_from_params
+    email = user_params.delete(:email)
+    username = user_params.delete(:username)
+    password = user_params.delete(:password)
+
+    Clearance.configuration.user_model.new(user_params).tap do |user|
+      user.email = email
+      user.username = username
+      user.password = password
+    end
```
Comment (georgebrock) [inline app/controllers/users_controller.rb]: I'm not sure why `email`, `username` and `password` are being assigned individually, while other fields are being assigned via `new(user_params)`. Is there something special about these three fields?

File: spec/models/user_spec.rb [hunk 12]
```diff
@@ -7,22 +7,17 @@
...
 
       expect(user.username).to eq "stardude440"
     end
 
     it "has an email" do
-      user = FactoryGirl.build_stubbed(:user)
+      user = FactoryGirl.build_stubbed(:user, email: "stardude440@example.com")
 
-      expect(user.email).to eq "stardude440@gmail.com"
+      expect(user.email).to eq "stardude440@example.com"
```
Comment (mike-burns) [inline spec/models/user_spec.rb]: :heart:

File: app/views/layouts/application.html.erb [hunk 13]
```diff
@@ -1,16 +1,28 @@
 <!DOCTYPE html>
 <html>
-  <head>
-    <title>MusicSnippets</title>
-    <%= csrf_meta_tags %>
-    <%= stylesheet_link_tag    'application', media: 'all', 'data-turbolinks-track': 'reload' %>
-    <%= javascript_include_tag 'application', 'data-turbolinks-track': 'reload' %>
-  </head>
+<head>
+  <%= javascript_include_tag 'application' %>
```
Comment (entcheva) [inline app/views/layouts/application.html.erb]: I was using single before I started at thoughtbot, but since double is the style here, I plan to convert to double. Will update as I go, and likely submit a dedicated PR at some point.

File: app/models/user.rb [hunk 14]
```diff
@@ -1,7 +1,9 @@
 class User < ApplicationRecord
-  has_secure_password
+  include Clearance::User
+
   has_many :snippets
   has_one :dashboard
-  validates :username, :email, uniqueness: true
+  
+  validates :email, uniqueness: true
```
Comment (entcheva) [inline app/models/user.rb]: Yes!

File: app/controllers/users_controller.rb [hunk 15]
```diff
@@ -16,7 +14,19 @@ def create
...
+  end
+
   def user_params
-    params.require(:user).permit(:username, :email, :password, :password_confirmation)
+    params[Clearance.configuration.user_parameter] || Hash.new
```
Comment (georgebrock) [inline app/controllers/users_controller.rb]: You could use `params.fetch(Clearance.configuration.user_parameter, Hash.new)` here.

If this was a normal `Hash` there wouldn't be much of a difference, but in this case `params` is an instance of [`ActionController::Parameters`](http://api.rubyonrails.org/classes/ActionController/Parameters.html). `ActionController::Parameters#fetch` will make sure any Hash you get back from `#fetch` is converted to an instance of `ActionController::Parameters`, which means you can call methods like `#require` and `#permit` on it.

File: db/migrate/20170509171615_remove_password_digest_from_user.rb (hunk 5 above)
Comment (entcheva) [inline db/migrate/20170509171615_remove_password_digest_from_user.rb]: Yep: [has_secure_password](http://api.rubyonrails.org/classes/ActiveModel/SecurePassword/ClassMethods.html#method-i-has_secure_password)

File: config/locales/clearance.en.yml (hunk 4 above)
Comment (entcheva) [inline config/locales/clearance.en.yml]: They are included in the gem, so we don't need to repeat them here.

File: .gitignore (hunk 8 above)
Comment (entcheva) [inline .gitignore]: 😬 👍 👍 👍

File: spec/models/snippet_spec.rb (hunk 6 above)
Comment (entcheva) [inline spec/models/snippet_spec.rb]: I like it. I'm using shoulda matchers in this spec on line 5, so maybe this test is redundant?

File: spec/models/user_spec.rb (hunk 7 above)
Comment (entcheva) [inline spec/models/user_spec.rb]: Are you thinking of a particular matcher?

File: .gitignore (hunk 8 above)
Comment (entcheva) [inline .gitignore]: I do have `.DS_Store`, etc. in my `.gitignore` but it decided to unignore. Will try again...

File: app/controllers/users_controller.rb (hunk 9 above)
Comment (entcheva) [inline app/controllers/users_controller.rb]: Awesome, thanks!

Comment (mike-burns) [inline app/controllers/users_controller.rb]: Whoops, I had a typo in the original (updated now): I was missing the `notice:` key to `redirect_to`.

File: config/routes.rb (hunk 10 above)
Comment (entcheva) [inline config/routes.rb]: Nope!

In this case the `UsersController` inherits from the controller inside Clearance, and we want to override that inheritance for just the `create` action, hence the explicit controller there. 

However, I removed it and it still works so 🙌

File: .gitignore (hunk 8 above)
Comment (Euraldius) [inline .gitignore]: You'll need to `git rm --cached .DS_Store` to tell git to no longer track the file, and then the `.gitignore` will take effect

File: app/models/snippet.rb [hunk 16]
```diff
@@ -1,5 +1,6 @@
 class Snippet < ApplicationRecord
   belongs_to :user
+  validates :name, :artist, :notes, presence: true
```
Comment (Euraldius) [inline app/models/snippet.rb]: Did you intend this validation to be part of this PR?

File: config/routes.rb [hunk 17]
```diff
@@ -1,11 +1,6 @@
...
   root to: 'dashboards#index', via: :get
   resource :dashboard, only: [:show]
-
-  get '/login', to: 'sessions#new'
-  post '/login', to: 'sessions#create'
-  get '/logout', to: 'sessions#destroy'
-
   resources :snippets, only: [:new, :create]
-  resources :users, only: [:new, :create]
+  resources :users, only: [:create]
```
Comment (Euraldius) [inline config/routes.rb]: Can we still reach our `UsersController` if we remove this line?

File: spec/features/user_logs_in_spec.rb [hunk 18]
```diff
@@ -4,13 +4,13 @@
   scenario "logs in" do
     user = FactoryGirl.create(:user)
 
-    visit login_path
+    visit sign_in_path
```
Comment (Euraldius) [inline spec/features/user_logs_in_spec.rb]: Can you use `rails generate clearance:specs` to generate clearance's feature specs and use those?

File: spec/models/snippet_spec.rb [hunk 19]
```diff
@@ -1,35 +1,14 @@
 require "rails_helper"
 
 RSpec.describe Snippet do
+
   describe "associations" do
-    it { is_expected.to belong_to(:user) }
+    it { is_expected.to belong_to :user }
```
Comment (Euraldius) [inline spec/models/snippet_spec.rb]: Nice use of shoulda :)

File: config/routes.rb (hunk 17 above)
Comment (entcheva) [inline config/routes.rb]: I get the following error when I remove line 5: `ActiveModel::ForbiddenAttributesError in Clearance::UsersController#create`

File: app/models/snippet.rb (hunk 16 above)
Comment (entcheva) [inline app/models/snippet.rb]: True, I think it makes more sense as part of this PR: [Add snippet feature spec #5](https://github.com/entcheva/music_snippets/pull/4)

File: spec/features/user_logs_in_spec.rb (hunk 18 above)
Comment (entcheva) [inline spec/features/user_logs_in_spec.rb]: 💡 Yup!
